regparser/api_writer.py

```diff
@@ -123,16 +131,396 @@ def write(self, python_object):
+
+        return replacement_offsets, replacement_texts
+
+    @staticmethod
+    def apply_paragraph_markers(text, replacements):
+
+        replacement_texts = []
+        replacement_offsets = []
+        for i, repl in enumerate(replacements):
+            marker_text = repl['text']
+            marker_length = len(marker_text)
+            marker_locations = repl['locations']
+            for loc in marker_locations:
+                offset = [loc, loc + marker_length]
+                replacement_offsets.append(offset)
+                replacement_texts.append('')
+
+        return replacement_offsets, replacement_texts
+
+    @staticmethod
+    def apply_internal_citations(text, replacements):
+
+        replacement_texts = []
+        replacement_offsets = []
+        for repl in replacements:
+            citation = repl['citation']
+            offsets = repl['offsets']
+            citation_target = '-'.join(citation)
+            for offset in offsets:
+                ref_text = text[offset[0]:offset[1]]
+                replacement_text = '<ref target="{}" reftype="internal">'.format(citation_target) + \
+                                   ref_text.encode('utf-8') + '</ref>'
+                replacement_offsets.append(offset)
+                replacement_texts.append(replacement_text)
+
+        return replacement_offsets, replacement_texts
+
+    @staticmethod
+    def apply_external_citations(text, replacements):
+
+        replacement_texts = []
+        replacement_offsets = []
+        for repl in replacements:
+            citation = map(str, repl['citation'])
+            citation_type = repl['citation_type']
+            offsets = repl['offsets']
+            for offset in offsets:
+                ref_text = text[offset[0]:offset[1]]
+                # we need to form a URL for the external citation based on the citation type
+                # I don't know how to do that yet so the target is just a placeholder
+                target_url = '{}:{}'.format(citation_type, '-'.join(citation))
+                replacement_text = '<ref target="{}" reftype="external">'.format(target_url) + \
+                                   ref_text.encode('utf-8') + '</ref>'
+                replacement_texts.append(replacement_text)
+                replacement_offsets.append(offset)
+
+        return replacement_offsets, replacement_texts
+
+    @staticmethod
+    def apply_definitions(text, replacement):
+
+        offset = replacement['offset']
+        term = replacement['term']
+        hash_value = hashlib.sha1(term).hexdigest()
+        replacement_text = text[offset[0]:offset[1]]
+        replacement_text = '<def term="{}" id="{}">'.format(term, hash_value) + \
+                           replacement_text.encode('utf-8') + '</def>'
+
+        return [offset], [replacement_text]
+
+
+    @staticmethod
+    def apply_keyterms():
+        pass
+
+    @staticmethod
+    def apply_formatting():
+        pass
+
```

Comment:
We could. The difference in my mind is that this concentrates specifically XML-related code in the part dedicated to handling XML specifically, but of course we could outsource this to the individual layers instead. I have to say that I am not a huge fan of the "general class - specific subclass with overrides" model; I find it somewhat impenetrable and creating multiple layers of abstraction that are hard to see through. In this case I think it would be ok though.